### PR TITLE
Corrected iTerm behaviour (don’t open new tab when no window is open)

### LINKED ISSUE #9: Pathfinder to iTerm not working
Using `PI` shortcut to open current Pathfinder location in iTerm is no longer functioning correctly.  It will open iTerm, however, it will not apply the expected path

### PR DESCRIPTION
**Pull request by vitorgalvao**
iTerm’s behaviour with this workflow is currently undesirable. When it runs, it *always* opens a new tab. This is desirable when we already have an iTerm window open, but undesirable when we don’t (since we’re left with an extra, useless tab, we have to manually close).

This is in part due to known difficulties of calling iTerm from Alfred with the correct behaviour.

@stuartcryan [devised a solution](https://github.com/stuartcryan/custom-iterm-applescripts-for-alfred), however. This PR grabs that solution, chops off what isn’t needed for this workflow, and modifies it a bit to be cosmetically more in line with the existing code.

This should also fix #9. The code in `pi` had multiple mistakes, and it needs nothing more than be a copy of `fi`. The only thing that needs to change is how to get the path, as the code to call iTerm is separate from that.

### CODE REVIEW

**Comment by LeEnno** [comment]
Huge thanks for your efforts. I will have a look as I soon as I find some time.

**Comment by LeEnno** [comment]
Sorry for the delay.

I honestly can't reproduce the bug you mentioned. But I know I had it before. Until now I thought I had fixed it but this seems to be not the case as you still experience it. May be some weird timing issue depending on execution speed, I don't know.

The disadvantage of your solution is that it doesn't consider users wo have iTerm with *no* windows open. But for the sake of simplicity let's just assume this won't apply to a lot of users.

Bottom line: I will just merge you pull request. Thank you a lot :)

**Comment by vitorgalvao** [comment]
> The disadvantage of your solution is that it doesn't consider users wo have iTerm with no windows open.

It does.

This solution is for iTerm 2.1.1, though, so future version might need an update. Check the linked repo, though, as it has the details (there’s also a script for iTerm 2.9).

**Comment by LeEnno** [comment]
> It does.

When closing all open windows in iTerm and using `fi`, it does not open a new window for me.

> This solution is for iTerm 2.1.1

Yeah, saw this. Might have a look into and try some kind of version condition.

**Comment by vitorgalvao** [comment]
> When closing all open windows in iTerm and using `fi`, it does not open a new window for me.

Seems like something might’ve changed. I can reproduce that as well, occasionally.

Code for 2.9 seems to be working well, though, so I’ve submitted that in #14.

**Comment by LeEnno** [comment]
> Seems like something might’ve changed. I can now reproduce that as well, occasionally.

Yes, initially I did all this if-else-code-pathing to ensure every case will be considered. I also thought the error you described initially in this issue would be caught as well, but as you reported it seemed to appear in your case.

The fix you pull requested didn't have all the edge cases in mind, so after merging it, the error I described happens *occasionally*.

Best solution will be to introduce some sort of testing for the case "iterm open without any windows" without getting rid of your solution. And I probably will do that. All just a matter of spare time in order to hack something together :)

**Comment by vitorgalvao** [comment]
> Best solution will be to introduce some sort of testing for the case "iterm open without any windows" without getting rid of your solution.

Problem is that isn’t simple for Alfred. There’s a known issue where Alfred and iTerm2 don’t work correctly for this exact case (of making a new tab or opening a new window). Testing for open windows and making tabs is very easy to have working in Applescript, the problem arises when that Applescript is inside Alfred.

That’s why @stuartcryan’s scripts are good and became popular, because they work for Alfred.

**Comment by LeEnno** [comment]
> That’s why @stuartcryan’s scripts are good and became popular, because they work for Alfred.

I thought after merging this PR we are currently using his scripts, aren't we?

**Comment by vitorgalvao** [comment]
Yes (although modified to be better in line with your code). But something changed in the meantime. From the time I submitted these to the time you merged it, his script was updated.

**Comment by LeEnno** [comment]
Ah, ok. I will have a look.

**Comment by vitorgalvao** [comment]
Here’s a preliminary try:

```applescript
on alfred_script(q)
  tell application "Finder" to set pathList to (quoted form of POSIX path of (folder of the front window as alias))
  tell application "System Events" to set isRunning to (exists (processes where name is "iTerm2")) or (exists (processes where name is "iTerm"))
  tell application "iTerm" to set iterm_version to version
  set iterm_version_minor to (do shell script "perl -pe 's/\\d\\.(\\d).*/\\1/ ' <<<" & iterm_version) as integer
  
  if (iterm_version_minor < 9) then
    if isRunning then
      tell application "iTerm"
        activate
        try
          set myterm to the first terminal
        on error
          set myterm to (make new terminal)
        end try
        tell myterm
          set mysession to (launch session "Default Session")
          tell mysession to write text "clear; cd " & pathList
        end tell
      end tell
    else
      tell application "iTerm"
        activate
        tell the first terminal
          tell the last session to write text "clear; cd " & pathList
        end tell
      end tell
    end if
  else
    if isRunning then
      tell application "iTerm"
        activate
        try
          select first window
          set onlywindow to false
        on error
          create window with default profile
          select first window
          set onlywindow to true
        end try
        tell the first window
          if onlywindow is false then
            create tab with default profile
          end if
          tell current session to write text "clear; cd " & pathList
        end tell
      end tell
    else
      tell application "iTerm"
        activate
        try
          select first window
        on error
          create window with default profile
          select first window
        end try
        tell the first window
          tell current session to write text "clear; cd " & pathList
        end tell
      end tell
    end if
  end if
end alfred_script
```

It isn’t working, though (a complaint in the 2.1 part of the code), but it is likely close. I’ll leave the rest to you, though.

To be honest, this is the type of workflow I don’t want to spend too much time on. If it didn’t exist, I’d do it just for my needs (only iTerm2, and only once version). This is pretty much feature-complete, and even adding support for Path Finder was already too much: It does exactly what I need and no new features are needed or will be added, which means only bug-fixing remains.

Keep all the code you want from this or close the PR, but personally I have no desire to spend the remainder of afternoon building and debugging something I’ll never use and will eventually get discarded (when 2.9 is stable).
